### DISCUSSION TITLE
How to add GTM in nextJS typescript (window.dataLayer is not a function)

### DISCUSSION THREAD

**Discussion by lucasReghiniM** [vercel/next.js#20784] [Help]
I am implementing GTM on my website, but my NextJS project is in Typescript.
I used the example of Github [the example of Github](https://github.com/vercel/next.js/tree/canary/examples/with-google-tag-manager), but this error appears:
**TypeError: window.dataLayer is not a function**
![image](https://user-images.githubusercontent.com/47827740/103675983-51ee6f00-4f5f-11eb-99ae-0bf2b728fe13.png)

This is my gtm.js file:

```
export const GTM_ID = process.env.TAG_MANAGER

export const pageview = (url) => {
  window.dataLayer({
    event: 'pageview',
    page: url,
  })
}
```
Yes, the GTM is in js, but all my project is in ts/tsx

My _app.tsx and my _document.tsx are the same as the example of vercel, i created the component and the .env with the ID

**Comment by gabeins** [comment]
In essence, `dataLayer` is just an array of objects, so you can add this to your type definitions:

```ts
declare global {
  interface Window {
    dataLayer: Record<string, any>[];
  }
}
```

**Reply by mikestopcontinues** [reply]
Hey, should this still work if I include it in `next-env.d.ts`? It only seems to be picked up when I place it in the same file as I reference `window.dataLayer`.

**Reply by gabeins** [reply]
@mikestopcontinues you either place it in the same file where you access `window.dataLayer`, or you have to add the file with your type definitions to the `types` array in your `tsconfig.json`:

```json
{
   "compilerOptions": {
       "types" : ["my-custom-type-definition-file.ts"]
   }
}
```

See: https://www.typescriptlang.org/tsconfig#types

⚠️ Be mindful that you will lose the automatic inclusion of `@types/*` in the global scope if you don't specify them.

> If types is specified, only packages listed will be included in the global scope. 

You can fix this by using `typeRoots` to define the path to your `node_modules/@types` folder: https://www.typescriptlang.org/tsconfig#typeRoots

**Reply by mikestopcontinues** [reply]
Thanks @gabrielizaias. Really helpful, especially on TS Day 1! :D

**Reply by yuriystasiv** [reply]
Maybe it will be useful for someone, I ended up with:

more info about export: https://stackoverflow.com/a/59499895/22062817
```
// custom.d.ts.
export {};

declare global {
  interface Window {
    dataLayer?: Record<string, unknown>[];
  }
}
```

```
// tsconfig.json
  "include": [
    "next-env.d.ts",
    "custom.d.ts",
    "**/*.ts",
    "**/*.tsx",
    ".next/types/**/*.ts"
  ],
```

**Comment by goliney** [comment]
I didn't want to modify `global` declaration, so I defined `window.dataLayer` in the only file that uses `dataLayer`:

```typescript
// gtm.ts

type WindowWithDataLayer = Window & {
  dataLayer: Record<string, any>[];
};

declare const window: WindowWithDataLayer;

export const pageview = (url: string) => {
  window.dataLayer.push({
    event: 'pageview',
    page: url,
  });
};
```

[TypeScript Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBA6glgOwCYHsDu9gAsAiBDYPAGTxAgCcoBeWRVNKAMigG8AoKKJA40igLigAlCAGMU5JAB4AzsHKIA5gBooeBCAB8AbQC6AbjYBfQ2yRiANnnLRxCOVDR10g+MnSZcPEmXKmIAB5gEsBQdg5geIoQAG5wEAw0ABQAruQWgnIKCIoAlNSarByOzmgAdNyEPhRlYCkyWEnsnJyxEAjAggDkkdFxCV3KxZy9EIJpFkOcRrmGJmxAA)

**Reply by ItsIrv** [reply]
Good idea! Can be simplified just a little bit more.

`declare const window: Window & {
  dataLayer: Record<string, unknown>[];
};`

**Comment by Elrinth** [comment]
This is probably unrelated to nextJS. but my client has google tag manager setup on their wordpress site (i checked source). I wrote a wordpress plugin using vue (typescript).

Trying to use the dataLayer global like this, unsure if it will work?
if ((window as any) && (window as any).dataLayer) {
          console.log("Sending dataLayer event"); // this is fired atleast

         // unsure if this does what it's supposed to do?!
          (window as any).dataLayer.push({
            event: "ajaxComplete",
            attributes: { type: "", etc: "" } });
           }
